File: src/commands/group/removemember.ts
import {ErrorCodes, GroupUserEditResponse, SDKError} from "@ironcorelabs/ironnode";
import {Command, flags as flagtype} from "@oclif/command";
import chalk from "chalk";
import * as GroupMaps from "../../lib/GroupMaps";
import {ironnode} from "../../lib/SDK";
import {keyFile, userList} from "../../lib/sharedFlags";
import {buildCommandSampleText, createDisplayTable} from "../../lib/Utils";

/**
 * Group remove member command. Takes a comma-separated list of emails as well as the group name and attempts to remove users as members from the group.
 */
export default class RemoveMember extends Command {
    static aliases = ["group:removemembers", "group:removeMember", "group:removeMembers"];
    static description = "Remove members from a group given their email address.";
    static args = [
        {
            name: "group",
            description: "Name of the group. Can alternately refer to a group by ID. Indicate IDs by prefixing with 'id^' e.g. 'id^groupID'.",
            required: true,
        },
    ];
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
        users: userList("Remove member permissions from the comma-separated list of user emails.")(),
    };
    static examples = [buildCommandSampleText("group:removemember -u test@example.com,test2@example.com myGroup")];

    /**
     * Build up a table that displays the results from the group remove member operation showing both successes and failures.
     */
    buildResultTable(buildResults: GroupUserEditResponse) {
        const table = createDisplayTable(["User", "Result"]);

        buildResults.succeeded.forEach((removedUser) => {
            table.push([removedUser, chalk.green("✔ Removed as member")]);
        });

        buildResults.failed.forEach((removedUser) => {
            table.push([removedUser.id, chalk.red(removedUser.error)]);
        });

        return table;
    }

    async run() {
        const {args, flags} = this.parse(RemoveMember);
        const groupID = await GroupMaps.getGroupIDFromName(args.group);

        return ironnode()
            .group.removeMembers(groupID, flags.users as string[])
            .then((removeResult) => {
                this.log(`\n${this.buildResultTable(removeResult).toString()}\n`);
            })
            .catch((e: SDKError) => {
                if (e.code === ErrorCodes.GROUP_REMOVE_MEMBERS_REQUEST_FAILURE) {
                    return this.error(chalk.red(`Unable to complete request to remove members from the group.`));
                }
                this.error(chalk.red(e.message));
            });
    }
}

File: src/group/remove_member.rs
use std::path::PathBuf;

use clap::Parser;
use ironoxide::{
    group::GroupAccessEditResult,
    prelude::{BlockingIronOxide, GroupId, GroupName, UserId},
    IronOxideErr,
};
use itertools::Either;

use crate::{
    group_maps::{self, convert_group_names_to_ids},
    util,
};

const EXAMPLE: &str = "EXAMPLE

    $ ironhide group remove-member -u test@example.com,test2@example.com myGroup

";

#[derive(Parser)]
/// Remove members from a group given their email address.
#[clap(after_help = EXAMPLE)]
pub struct RemoveMember {
    /// Name of the group. Can alternately refer to a group by ID. Indicate IDs by prefixing with 'id^' e.g. 'id^groupID'.
    #[clap(parse(try_from_str = util::group_identifier_from_string))]
    group: Either<GroupName, GroupId>,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(parse(from_os_str), short, long)]
    keyfile: Option<PathBuf>,
    /// Remove member permissions from the comma-separated list of user emails.
    #[clap(parse(try_from_str = util::try_from_email), short, long, use_value_delimiter = true, require_value_delimiter = true, required = true)]
    users: Vec<UserId>,
}

impl util::GetKeyfile for RemoveMember {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn group_remove_members(
    sdk: &BlockingIronOxide,
    RemoveMember { group, users, .. }: RemoveMember,
) -> Result<(), String> {
    let (groups_by_name, _) = group_maps::get_group_maps(sdk);
    let group_id = convert_group_names_to_ids(&[group], &groups_by_name)
        .first()
        .cloned()
        .expect("Group doesn't have an ID.");

    let removed_members = sdk.group_remove_members(&group_id, &users);
    let table = build_result_table(removed_members);
    table.printstd();

    Ok(())
}

fn build_result_table(
    remove_member_results: Result<GroupAccessEditResult, IronOxideErr>,
) -> prettytable::Table {
    let mut table = table!([Fbb->"User", Fbb->"Result"]);

    match remove_member_results {
        Ok(results) => {
            for succeeded_id in results.succeeded() {
                table.add_row(row![
                    Fw -> succeeded_id.id(),
                    Fg -> format!("{}", "✔ Removed as member"),
                ]);
            }
            for failure in results.failed() {
                table.add_row(row![
                    Fw -> failure.user().id(),
                    Fr -> failure.error(),
                ]);
            }
        }
        Err(e) => {
            table.add_row(
                row![Fw -> "ALL", Fr -> format!("Removing members failed catastrophically: {}", e)],
            );
        }
    }

    table
}
